cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -487,15 +489,27 @@ Status CopyFiles(const std::shared_ptr<FileSystem>& source_fs,
         internal::ConcatAbstractPath(destination_base_dir, relative->to_string());
 
     if (source_info.IsDirectory()) {
-      destination_dirs.insert(destination_path);
+      dirs.push_back(destination_path);
     } else if (source_info.IsFile()) {
       sources.push_back({source_fs, source_info.path()});
       destinations.push_back({destination_fs, destination_path});
     }
   }
 
-  std::vector<std::string> dirs(destination_dirs.size());
-  std::move(destination_dirs.begin(), destination_dirs.end(), dirs.begin());
+  // remove directories with descendants since recursive directory creation will create
+  // them automatically
+  std::sort(dirs.begin(), dirs.end());
+  for (auto ancestor = dirs.begin(); ancestor != dirs.end(); ++ancestor) {
+    auto descendants_end = ancestor + 1;
+
+    while (descendants_end != dirs.end() &&
+           internal::IsAncestorOf(*ancestor, *descendants_end)) {
+      ++descendants_end;
+    }
+
+    dirs.erase(ancestor, descendants_end - 1);
```

Comment:
Hmm... if I have `{"a", "a/b", "a/c")`, it looks like this will remove `{"a", "a/b"}`?

Reply:
no, `dirs.erase(ancestor, descendants_end)` would erase "a/b" too but I subtract 1 so it leaves the last descendant in place

Reply:
Perhaps `PathForest` can help. Though it's not obvious how to list the leaves.

Reply:
In this example, `descendants_end` would be `dirs.end()`, though?
Perhaps you can fix this by using `internal::IsAncestorOf(*(descendants_end - 1), *descendants_end)` instead of `internal::IsAncestorOf(*ancestor, *descendants_end)`.

Reply:
Another problem is that you're iterating on `dirs` while erasing from it. That's not safe, I think, unless you write `ancestor = dirs.erase(ancestor, descendants_end - 1)`.

Reply:
Yet another issue is if you have `{"a", "a/b/c"}`, `"a"` won't be removed?

Reply:
> Perhaps `PathForest` can help. Though it's not obvious how to list the leaves.
I don't think so; the intent of PathForest is to visit all ancestors efficiently so getting a minimal set of descendants would not be efficient

> `internal::IsAncestorOf(*(descendants_end - 1), *descendants_end)`
You're right, thanks

> Another problem is that you're iterating on dirs while erasing from it. That's not safe, I think, unless you write `ancestor = dirs.erase(ancestor, descendants_end - 1)`.
You're right; vector::erase invalidates iterators *at* or after the beginning of the range.

> Yet another issue is if you have {"a", "a/b/c"}, "a" won't be removed?
No, I'd expect `IsAncestorOf("a", "a/b/c") so "a" should be removed.